collector/src/main/java/io/prometheus/jmx/JmxScraper.java

```diff
@@ -93,82 +111,132 @@ public void doScrape() throws Exception {
-                    mBeanNames.add(instance.getObjectName());
+          
+          jmxc.addConnectionNotificationListener(new NotificationListener() {
+            @Override
+            public void handleNotification(Notification notification, Object handback) {
+                JMXConnectionNotification not = (JMXConnectionNotification) notification;
+                if (JMXConnectionNotification.CLOSED.equals(not.getType())
+                        || JMXConnectionNotification.FAILED.equals(not.getType())) {
+                    if (logger.isLoggable(Level.FINE)) logger.fine("JMX connection closed. Forcing reconnection.");
+                    try {
+                        jmxc.close();
+                    } catch (Exception e) {
+                        // Silence
+                    }
+                    jmxc = null;
+                    beanConn = null;
                 }
             }
+          }, null, null);
+        }
+    }
 
-            for (ObjectName name : blacklistObjectNames) {
-                for (ObjectInstance instance : beanConn.queryMBeans(name, null)) {
-                    mBeanNames.remove(instance.getObjectName());
-                }
-            }
+    /**
+      * If no connection available, (re)connects.
+      */
+    private void connectEventually() throws MalformedURLException, IOException {
+        if (jmxc == null) {
+            connect();
+        }
+    }
 
-            // Now that we have *only* the whitelisted mBeans, remove any old ones from the cache:
-            jmxMBeanPropertyCache.onlyKeepMBeans(mBeanNames);
+    /**
+      * Get a list of mbeans on host_port and scrape their values.
+      *
+      * Values are passed to the receiver in a single thread.
+      */
+    public void doScrape() throws Exception {
+        connectEventually();
+        // Query MBean names, see #89 for reasons queryMBeans() is used instead of queryNames()
+        Set<ObjectName> mBeanNames = new HashSet<ObjectName>();
+        for (ObjectName name : whitelistObjectNames) {
+            for (ObjectInstance instance : beanConn.queryMBeans(name, null)) {
+                mBeanNames.add(instance.getObjectName());
+            }
+        }
 
-            for (ObjectName objectName : mBeanNames) {
-                long start = System.nanoTime();
-                scrapeBean(beanConn, objectName);
-                logger.fine("TIME: " + (System.nanoTime() - start) + " ns for " + objectName.toString());
+        for (ObjectName name : blacklistObjectNames) {
+            for (ObjectInstance instance : beanConn.queryMBeans(name, null)) {
+                mBeanNames.remove(instance.getObjectName());
             }
-        } finally {
-          if (jmxc != null) {
-            jmxc.close();
-          }
+        }
+
+        // Now that we have *only* the whitelisted mBeans, remove any old ones from the cache:
+        jmxMBeanPropertyCache.onlyKeepMBeans(mBeanNames);
```

Comment:
Can we combine these caches?

Reply:
Hi.

Sorry for the delay, had no time these last days.
Good point. Done.